Rename Math to CompositeImplicitAutograd

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54470 Rename DefaultBackend to CompositeExplicitAutograd
* **#54466 Rename Math to CompositeImplicitAutograd**

I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27253239](https://our.internmc.facebook.com/intern/diff/D27253239)

ezyang committed:
Rename Math to CompositeImplicitAutograd

I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -18,15 +18,43 @@ struct OperatorKernel;
+//          This is good, because the CPU specific backend
+//          implementation is more specialized and typically better;
+//          if we used the composite, we would bypass it.
+//          (NB: the Autograd key is guaranteed to exists because
```
super nit: exist

bhosmer commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -18,15 +18,43 @@ struct OperatorKernel;
+//
+// As you can see, when we have a specific Autograd key (AutogradCPU), we can
+// decide whether or not to use the CompositeImplicitAutograd kernel or the
+// Autograd kernel based on whether or not the backend exists.
```
re "the backend exists" - should be "the backend kernel exists" I think, or "a backend kernel has been registered"

bhosmer reviewed (approved):
😍

ezyang committed:
Update on "Rename Math to CompositeImplicitAutograd"


I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Rename Math to CompositeImplicitAutograd"


I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Rename Math to CompositeImplicitAutograd"


I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Rename Math to CompositeImplicitAutograd"


I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27253239](https://our.internmc.facebook.com/intern/diff/D27253239)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/README.md:
```diff
@@ -264,64 +264,77 @@ dispatch:
+# out functions get suffixed with _out in their name
+func: func.out(...) -> ...
+dispatch:
+    CompositeImplicitAutograd: func_out
```
Hm, we don't support autograd for `out` functions, so this is pretty confusing

ezyang replied:
Hmm, I agree, and I'm not sure how to fix this problem.

ezyang replied:
Actually, I think this is OK. Let's take `dstack.out` as an example out function that has an implicit autograd composite registered for it. Its implementation looks like this:

```
Tensor& dstack_out(Tensor& result, TensorList tensors) {
  TORCH_CHECK(tensors.size() > 0,
           "dstack expects a non-empty TensorList");
  auto rep = at::atleast_3d(tensors);
  return at::cat_out(result, rep, 2);
}
```

This is fine; if `cat_out` had supported autograd, this would too. Similar to how composite functions support all backends their nested functions support, composite out functions support autograd if all their nested functions support autograd. OFC, right now they don't, but maybe this eventually won't be true.

ngimel replied:
Yeah, it's not a problem functionally, it's just confusing in the docs.

ngimel commented on aten/src/ATen/native/README.md:
```diff
@@ -264,64 +264,77 @@ dispatch:
+
+  - `CompositeExplicitAutograd` (previously known as `DefaultBackend`):
+    implementations of kernels that work for all backends, but require an
+    explicit definition of autograd in `derivatives.yaml` to support autograd.
```
`definition of backward function`?

ngimel commented on aten/src/ATen/native/README.md:
```diff
@@ -264,64 +264,77 @@ dispatch:
+    functions that do a very small amount of work and then delegate to another
+    operator to do the actual heavy lifting.  Under the hood, registering a
+    kernel to `CompositeExplicitAutograd` is equivalent to registering that
+    kernel to every backend (e.g., `CPU, CUDA`). Note: kernels which call
```
`Functions registered to DefaultBackend should work for any backend for inference.` was helpful in the previous docs, but probably with an added clarification of `if the nested functions that are called work for those backends`.

ngimel commented on aten/src/ATen/native/README.md:
```diff
@@ -264,64 +264,77 @@ dispatch:
+Whether to use implicit or explicit autograd for your kernel can be decided by the following steps:
+1. If you can, always start with a `CompositeImplicitAutograd` kernel that's composable from existing operators.
+2. If you don't want to use the derived gradient formula from `CompositeImplicitAutograd` kernel for autograd, either to
+   get better performance or better numerical stability, you should put the kernel in `CompositeExplicitAutograd`
```
you should register kernel to

ezyang committed:
Update on "Rename Math to CompositeImplicitAutograd"


I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27253239](https://our.internmc.facebook.com/intern/diff/D27253239)

[ghstack-poisoned]

ezyang committed:
Update on "Rename Math to CompositeImplicitAutograd"


I had to very carefully audit all the use sites since there are a lot
of other uses of the string Math; I did most of the conversion by
grepping for all occurrences of Math and then doing a search
replace.

I also updated documentation for clarity.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27253239](https://our.internmc.facebook.com/intern/diff/D27253239)

[ghstack-poisoned]

ailzhang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -408,7 +408,7 @@
   dispatch:
     CPU, CUDA: addr_out
-    Math: math_addr_out
+    CompositeImplicitAutograd: math_addr_out
```
I think all those _out Math kernels should've been added to DefaultBackend instead? It's likely  due to we don't have a way to test when a kernel claims it supports multiple backends. 
But indeed for _out it's fine to stay in Math since we probably want to consider support autograd for them in the future? cc: @albanD

ezyang replied:
Yeah, I think if they just call other at::_out functions, it's fine to keep them in Math. But really we need some strategy for testing that things are defined correctly

ailzhang reviewed (approved):
Thanks!

albanD reviewed (approved):
LGTM
